ateam_ai/src/trajectory_generation/ilqr_computer.hpp

```diff
@@ -0,0 +1,346 @@
+// Copyright 2021 A Team
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#ifndef TRAJECTORY_GENERATION__ILQR_COMPUTER_HPP_
+#define TRAJECTORY_GENERATION__ILQR_COMPUTER_HPP_
+
+#include "trajectory_generation/ilqr_problem.hpp"
+
+#include <Eigen/Dense>
+
+#include <optional>
+#include <tuple>
+#include <utility>
+
+struct iLQRParams
+{
+  // Max number of outer loop iterations
+  std::size_t max_ilqr_iterations = 1;
+  // Max number of forward pass iterations
+  std::size_t max_forward_pass_iterations = 1;
+  // Max change in the regulation on success/failure
+  double alpha_change = 0.5;
+  // Convergence threshold for stopping (delta cost)
+  double converge_threshold = 1e-1;
+};
+
+// Actually computes the calculations on a given iLQRProblem
+template<int X, int U, int T>
+class iLQRComputer
+{
+public:
+  using State = Eigen::Matrix<double, X, 1>;
+  using StateJacobian = Eigen::Matrix<double, X, X + U>;
+  using StateJacobian_x = Eigen::Matrix<double, X, X>;
+  using StateJacobian_x_T = Eigen::Matrix<double, X, X>;
+  using StateJacobian_u = Eigen::Matrix<double, X, U>;
+  using StateJacobian_u_T = Eigen::Matrix<double, U, X>;
+
+  using Input = Eigen::Matrix<double, U, 1>;
+
+  using Cost = double;
+
+  using CostGradiant = Eigen::Matrix<double, X + U, 1>;
+  using CostGradiant_x = Eigen::Matrix<double, X, 1>;
+  using CostGradiant_u = Eigen::Matrix<double, U, 1>;
+
+  using CostHessian = Eigen::Matrix<double, X + U, X + U>;
+  using CostHessian_xx = Eigen::Matrix<double, X, X>;
+  using CostHessian_ux = Eigen::Matrix<double, U, X>;
+  using CostHessian_uu = Eigen::Matrix<double, U, U>;
+
+  using Time = int;
+
+  using Trajectory = std::array<State, T>;
+  using Actions = std::array<Input, T>;
+
+  using Feedforwards = std::array<Input, T>;
+  using Feedbacks = std::array<Eigen::Matrix<double, U, X>, T>;
```

Comment:
C++ Question: Is there a way to share these aliases across classes? I tried having a common templated base class that both classes inherit from, but the compiler wasn't able to find the object types.